### PR TITLE
Optimize search for the default bundle

### PR DESCRIPTION
**Pull request by jiahaog**
`+[NSBundle bundleWithIdentifier]` is slow, and scales to the size of the app. This call during startup significantly affects the startup of `customer: chalk`. 

This implementation searches within the `privateFrameworksURL` where the default bundle is expected to be located. When tested internally, this improves startup of `customer: chalk` by 30% - see http://shortn/_WcifhO48X6.

It also falls back to `+[NSBundle bundleWithIdentifier]` so there is no functional change in behavior for customers that may be using alternate build systems.

*List which issues are fixed by this PR. You must list at least one issue.*

Fix flutter/flutter#37826

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by jmagman — COMMENTED** [review]
> When tested internally, this improves startup of `customer: chalk` by 30% - see http://shortn/_WcifhO48X6.

Wow, that's significant!

1. Does macOS have the same problem?
https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm#L29

2. Can you reuse this in other places in `FlutterDartProject`, see for example
https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L111-L112
https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L312

3. Testing
> As this is a pure performance change, no tests are added.

I don't think this is strictly true since it's adding lines of code that sure could have a bug in them, causing it to always incorrectly fall back to the slow implementation...
You could make `FLTBundleWithIdentifier` an `extern` in `FlutterDartProject_Internal.h` and test it explicitly. For example, you could create all the files needed to create a `NSBundle` in the test, pass the URL hint in, and then make sure that one is selected.

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm [hunk 1]
```diff
@@ -32,6 +32,37 @@
...
+// tens of milliseconds in a minimal flutter app, and closer to 100 milliseconds
+// in a medium sized Flutter app. It is likely that the slowness comes from
+// having to traverse and load all bundles known to the process. Using
+// `+[NSBundle allframeworks]` and filtering also suffers from the same problem.
+//
+// This implementation is an optimization to limit the search space with a hint.
+// Callers can provide a `hintURL` for where the bundle is expected to be
+// located in. If the desired bundle cannot be found here, the implementation
+// falls back to `+[NSBundle bundleWithIdentifier:]`.
+static NSBundle* FLTBundleWithIdentifier(NSString* bundleID, NSURL* hintURL) {
```
[the remark this answers is not in the document]
Comment (jiahaog) [reply shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm]: Done!

Also RE the general review comment:

> You could make `FLTBundleWithIdentifier` an `extern` in `FlutterDartProject_Internal.h` and test it explicitly.

This is pretty new to me, but it does seem like using an `extern` is incompatible with `NS_INLINE`, or `static` methods in general. I couldn't make `FLTFrameworkBundleInternal` static / extern, if we want to export it in `FlutterDartProject_Internal.h`. Please let me know if you have any ideas or if this is ok.

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm [hunk 2]
```diff
@@ -32,6 +32,37 @@
...
+  NSArray<NSURL*>* candidates = [[NSFileManager defaultManager]
```
[the remark this answers is not in the document]
Comment (jiahaog) [reply shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm]: I didn't know about this, thanks for the suggestion, done!

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm [hunk 3]
```diff
@@ -46,7 +77,11 @@
 
   bool hasExplicitBundle = bundle != nil;
   if (bundle == nil) {
-    bundle = [NSBundle bundleWithIdentifier:[FlutterDartProject defaultBundleIdentifier]];
+    // The default build system for Flutter places the default bundle in the
+    // same directory as the engine bundle (as they are both frameworks).
+    NSURL* defaultBundleHintURL = [engineBundle.bundleURL URLByDeletingLastPathComponent];
+    bundle =
+        FLTBundleWithIdentifier([FlutterDartProject defaultBundleIdentifier], defaultBundleHintURL);
```
[the remark this answers is not in the document]
Comment (jiahaog) [reply shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm]: Using this is actually much better and keeps the code simpler. It turns out that `NSBundle.mainBundle.privateFrameworksURL` resolves to the same thing as `engineBundle.bundleURL.URLByDeletingLastPathComponent`. As such, I removed the need to pass a hint and renamed `FLTBundleWithIdentifier` to `FLTFrameworkBundleWithIdentifier` and always search within this path.

**Comment by jiahaog** [comment]
Thank you for the in-depth review!

> 1. Does macOS have the same problem?
https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/macos/framework/Source/FlutterDartProject.mm#L29

I don't have a setup for profiling macOS apps now so I can't say for sure now, but I would expect this to also happen there. It seems like Chrome also encountered this problem on macOS - see `GetFrameworkBundlePath` (L156) of https://chromium.googlesource.com/chromium/src/+/master/chrome/common/chrome_paths_mac.mm.

> 2. Can you reuse this in other places in `FlutterDartProject`, see for example
> https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L111-L112

We don't have the source to `+[NSBundle bundleWithIdentifier:]` but at least empirically through profiling and experimentation, this method is slow because it has to iterate through all the known bundles and read the bundle IDs from the plist files. Since the bundle and path is already known here, it shouldn't be a problem here.

> https://github.com/flutter/engine/blob/706fd9cc2dd83e84018e6584f0fc595b8c489457/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L312

I changed this remaining call to `+[NSBundle bundleWithIdentifier:]` on iOS to use this implementation as well.

> You could make `FLTBundleWithIdentifier` an `extern` in `FlutterDartProject_Internal.h` and test it explicitly. For example, you could create all the files needed to create a `NSBundle` in the test, pass the URL hint in, and then make sure that one is selected.

I found it too hard to create a `NSBundle` in a test as this seems to be something that is created as part of the build process. Instead, I opted to try and find a known bundle that must / must not exist. I had to also factor out the implementation to not include the fallback to unit test the enumeration behavior specifically. Please take a look at the new changes in `FlutterDartProjectTest.mm` :)

**Review by jmagman — APPROVED** [review]
LGTM, thanks for tracking this weird perf issue down.

Filed https://github.com/flutter/flutter/issues/122028 to track sharing the optimization on macOS.

**Pull request merged**